test/extensions/filters/http/ext_proc/ext_proc_grpc_fuzz_persistent.cc

```diff
@@ -0,0 +1,133 @@
+#include "test/extensions/filters/http/ext_proc/ext_proc_grpc_fuzz.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace ExternalProcessing {
+
+using envoy::extensions::filters::http::ext_proc::v3::ProcessingMode;
+using envoy::service::ext_proc::v3::ProcessingRequest;
+using envoy::service::ext_proc::v3::ProcessingResponse;
+
+// Using persistent Envoy and ext_proc test server.
+static std::unique_ptr<ExtProcIntegrationFuzz> fuzzer = nullptr;
+static std::unique_ptr<ExtProcFuzzHelper> fuzz_helper = nullptr;
+// Protects fuzz_helper which will be accessed by multiple threads.
+static Thread::MutexBasicLockable fuzz_helper_lock_;
+
+static uint32_t fuzz_exec_count = 0;
+// Reset and recreate Envoy and test server after every 5000 fuzzer execs.
+#define FUZZ_CREATE_ENVOY(x) ((x) % 5000 == 0)
+
+DEFINE_PROTO_FUZZER(const test::extensions::filters::http::ext_proc::ExtProcGrpcTestCase& input) {
+  try {
+    TestUtility::validate(input);
+  } catch (const ProtoValidationException& e) {
+    ENVOY_LOG_MISC(debug, "ProtoValidationException: {}", e.what());
+    return;
+  }
+
+  // have separate data providers.
+  FuzzedDataProvider downstream_provider(
+      reinterpret_cast<const uint8_t*>(input.downstream_data().data()),
+      input.downstream_data().size());
+  FuzzedDataProvider ext_proc_provider(
+      reinterpret_cast<const uint8_t*>(input.ext_proc_data().data()), input.ext_proc_data().size());
+
+  // Initialize fuzzer once with IP and gRPC version from environment
+  if (FUZZ_CREATE_ENVOY(fuzz_exec_count)) {
+    fuzzer = std::make_unique<ExtProcIntegrationFuzz>(
+        TestEnvironment::getIpVersionsForTest()[0], TestEnvironment::getsGrpcVersionsForTest()[0]);
+  }
+  // Initialize fuzz_helper during every execution.
+  // This will be accessed by the test server which is initialized once.
+  fuzz_helper = std::make_unique<ExtProcFuzzHelper>(&ext_proc_provider);
+
+  // Initialize test server.
+  if (FUZZ_CREATE_ENVOY(fuzz_exec_count)) {
+    // This starts an external processor in a separate thread. This allows for the
+    // external process to consume messages in a loop without blocking the fuzz
+    // target from receiving the response.
+    fuzzer->test_processor_.start(
```

Comment:
It is challenging to review the diffs, but it seems this code repeats much of the previous code.
Can this be refactored differently (eliminating copy-paste, and ensuring that both the persistent and non-persistent fuzzers follow the same workflow)?

Reply:
As explained below, even though the test server runs the same work flow, however, the data it uses: fuzz_helper is very different.  In persistent case it is a global pointer, thus needs mutex protection when access it. In non-persistent case is a local variable.  I would like to keep them separated to keep them plain.

Reply:
I'm missing something here. This was also true for non-persistent variant, so what makes it different?